node/src/neighborhood/mod.rs

```diff
@@ -1314,58 +1510,121 @@ impl Neighborhood {
-                .filter(|node_record| {
-                    node_record.routes_data()
-                        || Self::is_orig_node_on_back_leg(**node_record, target_opt, direction)
-                })
-                .flat_map(|node_record| {
-                    let mut new_prefix = prefix.clone();
-                    new_prefix.push(node_record.public_key());
-
-                    let new_hops_remaining = if hops_remaining == 0 {
-                        0
-                    } else {
-                        hops_remaining - 1
-                    };
-
-                    let new_undesirability = self.compute_new_undesirability(
-                        node_record,
-                        undesirability,
-                        target_opt,
-                        new_hops_remaining,
-                        payload_size as u64,
-                        direction,
-                        hostname_opt,
-                    );
-
-                    self.routing_engine(
-                        new_prefix.clone(),
-                        new_undesirability,
-                        target_opt,
-                        new_hops_remaining,
-                        payload_size,
-                        direction,
-                        minimum_undesirability,
-                        hostname_opt,
-                    )
-                })
-                .collect()
+            self.routing_guts(
+                prefix,
+                undesirability,
+                target_opt,
+                hops_remaining,
+                payload_size,
+                direction,
+                minimum_undesirability,
+                hostname_opt,
+                research_neighborhood,
+                research_exits,
+                previous_node,
+            )
         }
     }
 
+    #[allow(clippy::too_many_arguments)]
+    fn routing_guts<'a>(
+        &'a self,
+        prefix: &mut [&'a PublicKey],
+        undesirability: i64,
+        target_opt: Option<&'a PublicKey>,
+        hops_remaining: usize,
+        payload_size: usize,
+        direction: RouteDirection,
+        minimum_undesirability: &mut i64,
+        hostname_opt: Option<&str>,
+        research_neighborhood: bool,
+        research_exits: &mut Vec<&'a PublicKey>,
+        previous_node: &NodeRecord,
+    ) -> Vec<ComputedRouteSegment> {
+        // Go through all the neighbors and compute shorter routes through all the ones we're not already using.
+        previous_node
+            .full_neighbors(&self.neighborhood_database)
+            .iter()
+            .filter(|node_record| !prefix.contains(&node_record.public_key()))
+            .filter(|node_record| {
+                node_record.routes_data()
+                    || Self::is_orig_node_on_back_leg(**node_record, target_opt, direction)
+            })
+            .flat_map(|node_record| {
+                let mut new_prefix = prefix.to_owned();
```

Comment:
Tady se nejspis muzes vratit k tomu staremu zpusobu s clone(). `.to_owned()` mimochodem dela ve vysledku to same: Vezme objekt s referenci a udela novy, identicky, ale bez reference. To nejde udelat bez klonovani.